controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+			vmList = append(vmList, vm)
+		}
+
+		if len(errs) > 0 {
+			return apierrorsutil.NewAggregate(errs)
+		}
+
+		return r.waitForMachineCreation(ctx, vmList)
+	case diff > 0:
+		ctx.Logger.Info("ReplicaSet is scaling down",
+			"currentReplicas", len(vms),
+			"desiredReplicas", *(rs.Spec.Replicas),
+			"vmsToBeCreated", diff,
+			"deletePolicy", "oldestFirst",
+		)
+
+		deletePriorityFunc, err := getDeletePriorityFunc(rs)
+		if err != nil {
+			return err
+		}
+
+		var errs []error
+		vmsToDelete := getMachinesToDeletePrioritized(vms, diff, deletePriorityFunc)
+		for i, vm := range vmsToDelete {
+			log := ctx.Logger.WithValues("vm", vm.Name)
+			if vm.GetDeletionTimestamp().IsZero() {
+				log.Info("Deleting VM to scale down replicaset", "index", i+1, "totalVMsToBeDeleted", diff)
+
+				if err := r.Client.Delete(ctx, vm); err != nil {
+					log.Error(err, "Unable to delete VM")
+					r.Recorder.Warnf(rs, "FailedDelete", "Failed to delete VM %q: %v", vm.Name, err)
+					errs = append(errs, err)
+					continue
+				}
+				r.Recorder.Eventf(rs, "SuccessfulDelete", "Deleted VM %q", vm.Name)
+			} else {
+				log.Info(fmt.Sprintf("Waiting for VM %d of %d to be deleted", i+1, diff))
+			}
+		}
+
+		if len(errs) > 0 {
+			return apierrorsutil.NewAggregate(errs)
+		}
+		return r.waitForMachineDeletion(ctx, vmsToDelete)
 	}
 
-	ctx.Logger.Info("Finished Reconciling VirtualMachineReplicaSet Deletion")
+	return nil
+}
+
+func (r *Reconciler) waitForMachineDeletion(ctx *pkgctx.VirtualMachineReplicaSetContext, vmList []*vmopv1.VirtualMachine) error {
+	for _, vm := range vmList {
+		pollErr := wait.PollUntilContextTimeout(ctx, stateConfirmationInterval, stateConfirmationTimeout, false, func(ctx context.Context) (bool, error) {
+			m := &vmopv1.VirtualMachine{}
+			key := client.ObjectKey{Namespace: vm.Namespace, Name: vm.Name}
+			err := r.Client.Get(ctx, key, m)
+			if apierrors.IsNotFound(err) || !m.DeletionTimestamp.IsZero() {
+				return true, nil
+			}
+			return false, err
+		})
+
+		if pollErr != nil {
+			ctx.Logger.Error(pollErr, "Failed waiting for VirtualMachine to be deleted")
+			return errors.Wrap(pollErr, "failed waiting for VirtualMachine to be deleted")
+		}
+	}
+	return nil
+}
+
+func (r *Reconciler) waitForMachineCreation(ctx *pkgctx.VirtualMachineReplicaSetContext, vmList []*vmopv1.VirtualMachine) error {
+	for _, vm := range vmList {
+		pollErr := wait.PollUntilContextTimeout(ctx, stateConfirmationInterval, stateConfirmationTimeout, false, func(ctx context.Context) (bool, error) {
+			key := client.ObjectKey{Namespace: vm.Namespace, Name: vm.Name}
+			if err := r.Client.Get(ctx, key, &vmopv1.VirtualMachine{}); err != nil {
+				if apierrors.IsNotFound(err) {
+					return false, nil
+				}
+				return false, err
+			}
+
+			return true, nil
+		})
+
+		if pollErr != nil {
+			ctx.Logger.Error(pollErr, "Failed waiting for VirtualMachine to be created")
+			return errors.Wrap(pollErr, "failed waiting for VirtualMachine to be created")
```

Comment:
We are not using `github.com/pkg/errors` anymore. Please change this to:

```go
fmt.Errorf("failed waiting for VirtualMachine to be created: %w", err)
```

Please see https://github.com/vmware-tanzu/vm-operator/pull/538 for more information.